debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/AnalyzedBatchFailedWithBulkWriteException.java

```diff
@@ -0,0 +1,133 @@
+                            new AbstractMap.SimpleImmutableEntry<>(
+                                    batch.get(writeErrorIdx), new WriteException(writeError)));
+                });
+        if (ordered && !writeErrors.isEmpty()) {
+            Assertions.assertTrue(writeErrors.size() == 1);
+            int writeErrorIdx = writeErrors.get(0).getIndex();
+            for (int i = writeErrorIdx + 1; i < batch.size(); i++) {
+                skippedRecords.put(i, batch.get(i));
+            }
+        }
+        if (writeConcernError != null) {
+            IntStream.range(0, batch.size())
+                    .filter(
+                            i -> !recordsFailedWithWriteError.containsKey(i) && !skippedRecords.containsKey(i))
+                    .forEach(i -> recordsFailedWithWriteConcernError.put(i, batch.get(i)));
+        }
+    }
+
+    public void log() {
+        if (!recordsFailedWithWriteError.isEmpty()) {
+            recordsFailedWithWriteError.forEach(
+                    (i, recordFailedWithWriteError) -> logger.log(
+                            singletonList(recordFailedWithWriteError.getKey()),
+                            recordFailedWithWriteError.getValue()));
+        }
+        if (!recordsFailedWithWriteConcernError.isEmpty()) {
+            logger.log(recordsFailedWithWriteConcernError.values(), Assertions.assertNotNull(writeConcernException));
+        }
+        if (!skippedRecords.isEmpty()) {
+            logger.log(skippedRecords.values(), writeSkippedException);
+        }
+    }
+
+    /**
+     * {@linkplain #errorReporter Reports} records in the batch that either definitely have not been
+     * written, or may not have been written. While it is unclear whether the order of reported
+     * records in the DLQ is the same as the order in which they are reported, we still report them in
+     * the order they are present in the batch just in case because it is trivial to do.
+     */
+    public void report() {
+        for (int i = 0; i < batch.size(); i++) {
+            Map.Entry<SinkRecord, WriteException> recordFailedWithWriteError = recordsFailedWithWriteError.get(i);
+            if (recordFailedWithWriteError != null) {
+                errorReporter.report(
+                        recordFailedWithWriteError.getKey(), recordFailedWithWriteError.getValue());
+                continue;
+            }
+            SinkRecord recordFailedWithWriteConcernError = recordsFailedWithWriteConcernError.get(i);
+            if (recordFailedWithWriteConcernError != null) {
+                errorReporter.report(
+                        recordFailedWithWriteConcernError, Assertions.assertNotNull(writeConcernException));
+                continue;
+            }
+            SinkRecord skippedRecord = skippedRecords.get(i);
+            if (skippedRecord != null) {
+                errorReporter.report(skippedRecord, writeSkippedException);
+            }
+        }
+    }
+
+    public interface Logger {
```

Comment:
Could we have a different/more specific name? `Logger` is too associated with log4j.

Reply:
Removed